samples: nrf9160: mqtt_simple: Add extended TLS support

simensrostad opened the pull request:
Add certificate provisioning, sec tag socket configuration
and TLS session caching option.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## Nits issues

```

Redundant 'source "$(ZEPHYR_BASE)/Kconfig.zephyr" in 'samples/nrf9160/mqtt_simple/Kconfig'. Just do 'source "Kconfig.zephyr"'
instead. The $srctree environment variable already points to the Zephyr root,
and all 'source's are relative to it.
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

sigvartmh commented on samples/nrf9160/mqtt_simple/Kconfig:
```diff
@@ -33,8 +33,19 @@ config MQTT_PAYLOAD_BUFFER_SIZE
 
 menu "Zephyr Kernel"
-source "$ZEPHYR_BASE/Kconfig.zephyr"
+source "Kconfig.zephyr"
```
Why this change? I don't think it should be in here.

simensrostad replied:
The NordicBuilder suggested this change.

sigvartmh replied:
Then I guess it's fine :+1: just looked like a weird change. I guess it's because the Kconfig of the sample is quite old.

sigvartmh reviewed (approved):
LGTM

rlubos commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -332,10 +381,39 @@ static void client_init(struct mqtt_client *client)
+					    TLS_SESSION_CACHE_ENABLED :
+					    TLS_SESSION_CACHE_DISABLED;
+
+	tls_cfg->peer_verify = CONFIG_MQTT_TLS_PEER_VERIFY;
```
Isn't it needed to set these parameters for qemu build as well? Right now running the sample for qemu will skip the certificate verification if I understand correctly.

simensrostad replied:
Yes, you are correct.

rlubos commented on samples/nrf9160/mqtt_simple/src/certificates.h:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#define CLIENT_PRIVATE_KEY \
```
Do we really need to authenticate the client in this simple sample? Wouldn't server certificate verification be enough? I'm not even sure if the default server (mqtt.eclipse.org) allows that.

simensrostad replied:
Yes, the server certificate should be enough. I have added the CA cert for "test.mosquitto.org".

rlubos reviewed (commented):
What server was used for testing? Perhaps we should provide the certificate for the default server (mqtt.eclipse.org, from what I've seen it supports TLS)., so that the sample can be run with TLS out of the box? We should also add an overlay file to enable TLS for nRF91 (and possibly mentions TLS support in the sample documentation).

simensrostad commented:
@rlubos I have addressed your review comments. I have done some testing and the sample now works out of the box with no TLS/TLS for both qemu and the nrf91.

rlubos commented on samples/nrf9160/mqtt_simple/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018 Nordic Semiconductor
+# Copyright (c) 2020 Nordic Semiconductor
```
There's no need to update the year here, there were no modifications in this file.

rlubos commented on samples/nrf9160/mqtt_simple/overlay-carrier.conf:
```diff
@@ -1,8 +1,6 @@
-#
```
Let's keep the copyright notice.

rlubos commented on samples/nrf9160/mqtt_simple/overlay-qemu-x86-tls.conf:
```diff
@@ -0,0 +1,21 @@
+# To build the mqtt simple sample with TLS support
```
I think this information should rather be placed in the README.rst, let's add the copyright info instead.

rlubos commented on samples/nrf9160/mqtt_simple/overlay-tls.conf:
```diff
@@ -0,0 +1,8 @@
+# To build the mqtt simple sample with TLS support
```
Same story here, please add copyright notice.

The README.rst has already a section on how to add TLS support to the sample manually, which IMO is no longer required after this update. Perhaps we should replace it with information about overlays:
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/nrf9160/mqtt_simple/README.rst#references

simensrostad replied:
Sound good. I will update the documentation to reflect the changes that have been made, and remove/rephrase how the external link regarding TLS support is added. It makes sense to link to the external TLS doc for potential issues related to TLS. If people have issues they can post comments there.

rlubos commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -332,10 +368,38 @@ static void client_init(struct mqtt_client *client)
+	/* TLS session caching is not supported by the Zephyr network stack */
+	tls_cfg->session_cache = TLS_SESSION_CACHE_DISABLED;
+
+	err = certificates_provision();
```
If I understand correctly, the user will need to provision certificate to the modem manually, before running the sample. Perhaps we could automate the process, just as `https_client` sample does, and pre-load the certificate into the modem in the sample?
https://github.com/nrfconnect/sdk-nrf/blob/master/samples/nrf9160/https_client/src/main.c#L171
Note, that certificate should be loaded before `modem_configure` (autoconnect is already disabled in this sample).

simensrostad replied:
You are correct. Your suggestion sounds like a fine improvement. I will add certificate provisioning for the nrf91.

rlubos reviewed (commented):
Looks pretty good, just some suggestions to make the process simpler for the users.

jtguggedal commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -395,7 +459,10 @@ void main(void)
-	client_init(&client);
+	err = client_init(&client);
+	if (err != 0) {
+		printk("ERROR: client_init %d\n", err);
```
It looks like an error here could cause the subsequent functions to fail (depending on peer verification settings) so it should be considered to return.

Slightly besides the scope of this PR: I think it should also be considered to let `broker_init()` return an error and propagate it from `client_init()`. Error handling in `broker_init()` looks incomplete, and failure in that function would cause the connection attempt to fail.

simensrostad replied:
Let's add this while we are at it.

jtguggedal commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -34,6 +38,31 @@ static bool connected;
+		return err;
+	}
+
+	certs_added = true;
```
Looks like this function is only called once, so the flag is sort of redundant in this use-case

simensrostad replied:
Removed the flag

jtguggedal commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -10,6 +10,10 @@
 #include <string.h>
 #include <random/rand32.h>
 
+#if defined(CONFIG_BOARD_QEMU_X86) && !defined(CONFIG_BSD_LIBRARY)
```
Let's include this file for all builds, and offer certificate provisioning functionality also for builds using BSDlib

jtguggedal commented on samples/nrf9160/mqtt_simple/overlay-tls.conf:
```diff
@@ -0,0 +1,8 @@
+
+# MQTT TLS
+CONFIG_MQTT_LIB_TLS=y
+CONFIG_MQTT_BROKER_HOSTNAME="test.mosquitto.org"
```
I tested that it's also possible to connect to the default broker `mqtt.eclipse.org`

simensrostad replied:
Very good, without changing the CA?

simensrostad replied:
Changed the default broker to mqtt.eclipse.org and added the corresponding CA.

jtguggedal reviewed (commented):
Overall looks good! Left a few comments to consider.

simensrostad commented:
@umapraseeda I will update the documentation for this sample. It will need a documentation review, thanks.

jtguggedal commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -305,15 +346,22 @@ static void broker_init(void)
+	err = broker_init();
+	if (err) {
+		printk("Failed to initialize broker connection\n");
+	}
```
Error should be returned here

simensrostad replied:
woops, added

jtguggedal commented on samples/nrf9160/mqtt_simple/src/main.c:
```diff
@@ -266,12 +310,10 @@ static void broker_init(void)
 		printk("ERROR: getaddrinfo failed %d\n", err);
-
-		return;
+		return err;
```
Note that `getaddrinfo` returns positive error codes

simensrostad replied:
Changed to -err;

jtguggedal replied:
Okay, that's confusing. Better translate it to a negative error code than (as I see you've already done)

rlubos reviewed (approved):
LGTM, thanks!

jtguggedal reviewed (approved):
LGTM
